Improve JSON support; move most ser/deser from generated code into runtime library

garyp opened the pull request:
* Rewrote JSON support to better implement the proto3 JSON spec. This includes support for the `json_name` protobuf option.

* Added `JsonConfig` class that can be used for configuring JSON marshalling/unmarshalling at runtime. Currently supported configuration includes `ignoreUnknownFieldsInInput`, `outputDefaultValues`, and `outputProtoFieldNames`, which match the options described at https://developers.google.com/protocol-buffers/docs/proto3#json_options.

* Got rid of the need for projects using `pbandk` to use the `kotlinx-serialization` gradle plugin and add a dependency on the `kotlinx-serialization` library. The library is now an internal implementation detail of `pbandk`.

* Moved all of the binary marshalling functionality and most of the binary unmarshalling functionality from the generated code to the runtime library.

* Added `protoMarshal(OutputStream)`, `protoUnmarshal(InputStream)`, and `protoUnmarshal(ByteBuffer)` overloads on the JVM.

* Fixed some bugs with handling of `packed` fields, marshalling of enums in Javascript, and base64 encoding/decoding in Javascript.

Breaking API changes:

* Changed `Message` so that it no longer needs to extend from itself.

* Added `@PbandkInternal` and `@PublicForGeneratedCode` annotations on relevant portions of the public API that are only public for pbandk's internal use.

* The below methods are now defined as extension methods rather than being part of the `Message` interface. Code that calls these methods will now need to import them first.

    * `Message.jsonMarshal()`
    * `Message.Companion.jsonUnmarshal(String)`
    * `Message.protoMarshal()`
    * `Message.Companion.protoUnmarshal(ByteArray)`

* These two method overloads:

    * `Message.jsonMarshal(kotlinx.serialization.json.Json)`
    * `Message.Companion.jsonUnmarshal(kotlinx.serialization.json.Json, String)`

    have been replaced with:

    * `Message.jsonMarshal(pbandk.json.JsonConfig)`
    * `Message.Companion.jsonUnmarshal(String, pbandk.json.JsonConfig)`

* These two method overloads:

    * `Message.protoMarshal(pbandk.Marshaller)`
    * `Message.Companion.protoUnmarshal(pbandk.Unmarshaller)`

    have been replaced with:

    * `Message.marshal(pbandk.MessageMarshaller)`
    * `Message.Companion.unmarshal(pbandk.MessageUnmarshaller)`

* Replaced `Marshaller` and `Unmarshaller` interfaces with `MessageMarshaller` and `MessageUnmarshaller`, which are much simpler and function differently from the previous interfaces.

* Removed `Sizer` and `Util` from the public API.

* Removed `UnknownField` constructors and the `UnknownField.size()` method from the public API.

* `MessageMap.entries` is now a `Set` instead of a `List`.

Fixes #23, fixes #61. Portions of the proto3 JSON spec that still need to be implemented are tracked in #72.

seanadkinson commented on runtime/src/commonMain/kotlin/pbandk/json/JsonWireMarshaller.kt:
```diff
@@ -0,0 +1,183 @@
+            val value = (fd.value as KProperty1<T, *>).get(message)
+
+            if (value == null && fd.oneofMember) continue
+            if (!fd.oneofMember && !jsonConfig.outputDefaultValues && fd.type.isDefaultValue(value)) continue
```
Does this mean we won't have empty strings in our response now? 🎉

garyp replied:
Yeah. By default (when `jsonConfig.outputDefaultValues == false`) empty strings, zero values, false values, and empty lists will be omitted from the response.

One idea for a future enhancement is to add a setting along the lines of `jsonConfig.outputDefaultValuesForFields = fieldList` to allow you to choose specific fields whose default values will be output. I think we had talked about something like this in the past.

seanadkinson replied:
Yeah, we previously explicitly allowed `total_count` to return as `0`, which was useful.

seanadkinson commented on runtime/src/commonMain/kotlin/pbandk/json/JsonWireMarshaller.kt:
```diff
@@ -0,0 +1,183 @@
+//
+//    fun writeSInt32(value: Int): JsonElement = JsonLiteral(value)
+//
+//    fun writeSInt64(value: Long): JsonElement = JsonLiteral(value.toString())
```
Why commented out?

garyp replied:
This is some of the cleanup I mentioned I still need to do. These are currently all going through the generic `writeNumber32` and `writeNumber64` methods instead, since JSON uses the same serialization for all 32-bit and all 64-bit number types. Though I'm considering bringing these methods back and having `JsonValueMarshaller` extend from the `ValueMarshaller` interface (it doesn't currently), which I think will enable some useful behavior in the future.

seanadkinson commented on runtime/src/commonMain/kotlin/pbandk/json/JsonWireUnmarshaller.kt:
```diff
@@ -0,0 +1,264 @@
+        throw InvalidProtocolBufferException("unable to read message", e)
+    }
+
+    private fun <T : Message> readMessageObject(
+        messageCompanion: Message.Companion<T>,
+        content: JsonElement,
+        fieldFn: (Int, Any) -> Unit
+    ) {
+        for ((key, jsonValue) in content.jsonObject) {
+            val fd = messageCompanion.fieldDescriptors.firstOrNull { key in it.jsonNames }
+                ?: if (jsonConfig.ignoreUnknownFieldsInInput) {
+                    continue
+                } else {
+                    throw InvalidProtocolBufferException("Unknown field name and ignoreUnknownFieldsInInput=false: $key")
+                }
+            if (jsonValue.isNull) continue
+            fieldFn(fd.number, jsonValueUnmarshaller.readValue(jsonValue, fd.type))
+        }
+    }
```
So much nicer!

seanadkinson commented on runtime/src/commonMain/kotlin/pbandk/json/MessageExtensions.kt:
```diff
@@ -0,0 +1,9 @@
+    JsonWireMarshaller(jsonConfig).also{ it.writeMessage(this) }.toJsonString()
+
+fun <T : Message> Message.Companion<T>.jsonUnmarshal(data: String, jsonConfig: JsonConfig = JsonConfig.DEFAULT): T =
+    unmarshal(JsonWireUnmarshaller.fromString(data, jsonConfig))
```
Nice to put these extensions only once, as opposed to each generated file

garyp replied:
I also like that all of the JSON-related functionality is segregated into separate files. I'm considering in the future splitting these out into a separate `pbandk-json` library for folks that don't use JSON and don't want the extra code and dependencies needed for JSON support.

garyp commented:
A couple things I meant to mention:

1. This PR is still creating an `unmarshal()` method in each generated class that has to know about the structure of the message. That's temporary. Once we add support for message builders, we can create a generic version of `unmarshal()` that's driven entirely off of the `Message.fieldDescriptors` list, the same way that this PR is doing for marshalling. I'm pretty excited about that because it'll really slim down the generated classes without a big runtime performance impact.

2. This PR only converts the `jsonMarshal()` and `jsonUnmarshal()` methods over to these new `WireMarshaller` and `WireUnmarshaller` interfaces. My intent is to eventually rewrite the current `protoMarshal()` and `protoUnmarshal()` implementations to use these new interfaces as well. That will centralize all knowledge about (de)ser into the runtime library and make the generated code completely unaware about the serialization format.

garyp commented:
@seanadkinson This is finally ready for review. One small fix led to another and I ended up converting all of the binary marshall/unmarshall code as well 😁 The new code happens to fix some additional conformance tests we were previously failing. I'll create GitHub issues for the remaining gaps in our JSON support.

seanadkinson commented on protoc-gen-kotlin/lib/src/commonMain/kotlin/pbandk/gen/CodeGenerator.kt:
```diff
@@ -190,7 +153,7 @@ open class CodeGenerator(val file: File, val kotlinTypeMappings: Map<String, Str
                 line("${field.kotlinFieldName} = " +
                     "${field.kotlinFieldName}?.plus(plus.${field.kotlinFieldName}) ?: plus.${field.kotlinFieldName},")
-            } else if (file.version == 2 && field.optional) {
+            } else if (field.hasPresence) {
```
`hasPresence` means that its optional?

garyp replied:
Yeah. This is the new terminology that protobuf is using now that they added `optional` support in proto3. If you're curious, this has more details: https://github.com/protocolbuffers/protobuf/blob/master/docs/field_presence.md.

seanadkinson commented on runtime/src/commonMain/kotlin/pbandk/internal/binary/BinaryMessageMarshaller.kt:
```diff
@@ -0,0 +1,120 @@
+        sizeFn: (T) -> Int
+    ) {
+        val valueType = type.messageCompanion.fieldDescriptors[0].type
+        if (valueType.isDefaultValue(value)) {
```
Theres another default value check up on line 14.  Is one of these not needed?

garyp replied:
They're slightly different for wrapper types. Using the `StringValue` wrapper type as an example:
* The call on line 14 is doing `!fd.type.hasPresence && fd.type.isDefaultValue(value)`. `StringValue` is a message type and message types always "have presence" (i.e. they're nullable), so the `fd.type.isDefaultValue(value)` check won't even be called on line 14.
* Here on line 62, we've looked up the underlying type of the `StringValue` wrapper: `String`. We're now checking if `value` is the default value of `String` (not `StringValue`). If it is, we'll omit writing the empty string on the wire. But we're still writing the `StringValue` message header so the receiver knows that the value was non-null.

seanadkinson commented on runtime/src/commonMain/kotlin/pbandk/internal/json/JsonMessageMarshaller.kt:
```diff
@@ -0,0 +1,66 @@
+            // All other message types
+            else -> writeMessageObject(message)
+        }
+    }
```
So simple 🎉

seanadkinson commented on runtime/src/jsMain/kotlin/pbandk/internal/binary/ProtobufjsBinaryWireMarshaller.kt:
```diff
@@ -0,0 +1,133 @@
+import pbandk.protobufjs.Writer
+import pbandk.protobufjs.protobufjsLong
+
+private fun Writer.writeValueNoTag(type: FieldDescriptor.Type, value: Any) {
```
Is this code unable to use the `JsonValueMarshaller` and related classes from `commonMain`?  I guess I'm unclear about the relation of this class (and `CodedStreamBinaryWireMarshaller` below) to the ones in `commonMain`.

garyp replied:
I think you're confusing the JSON marshallers and the "binary" marshallers. `ProtobufjsBinaryWireMarshaller` and `CodedStreamBinaryWireMarshaller` are both responsible for converting a field into protobuf binary format. Whereas `JsonValueMarshaller` is responsible for converting a field into protobuf JSON format.

seanadkinson replied:
Yes, sorry, I understand the different between the JSON and the Binary - bad example specifying `JsonValueMarshaller` in my comment.

The point I was trying to make was that it seemed like there was a lot of duplicate work being done in this `jsMain` module, that is already done in `commonMain`, but looking closer, this probably isn't the case.  I guess I just saw enough `when` statements that rolled through each type, either reading or writing the value, and I thought we wouldn't need to do that outside of `commonMain`.  But if I understand correctly, this code is just providing the javascript bindings of doing the actual reading/writing, and code up in `commonMain` is using this marshaller for doing the work in javascript.  Is my understanding correct?

garyp commented:
Now that the Kotlin/Native PR finally got merged, I've rebased on top of that one. I've also added a detailed description to this PR with all of the changes being made. I'll merge this as soon as the tests pass.
